client/asset/dcr/externaltx.go

```diff
@@ -0,0 +1,444 @@
+}
+
+func (output *externalTxOutput) String() string {
+	return fmt.Sprintf("%s:%d", output.txHash, output.vout)
+}
+
+type blockFiltersScanner struct {
+	scanMtx          sync.RWMutex
+	lastScannedBlock *chainhash.Hash
+	relevantBlock    *block
+}
+
+// relevantBlockHash returns the hash of the block relevant to this scanner, if
+// the relevantBlok is set and is a mainchain block. Returns a nil hash and nil
+// error if the relevantBlock is set but is no longer part of the mainchain.
+// The scanner's scanMtx MUST be write-locked.
+func (scanner *blockFiltersScanner) relevantBlockHash(nodeGetBlockHashFn func(int64) (*chainhash.Hash, error)) (*chainhash.Hash, error) {
+	if scanner.relevantBlock == nil {
+		return nil, nil
+	}
+	mainchainBlockHash, err := nodeGetBlockHashFn(scanner.relevantBlock.height)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get hash for block %d: %v", scanner.relevantBlock.height, err)
+	}
+	if mainchainBlockHash.IsEqual(scanner.relevantBlock.hash) {
+		return scanner.relevantBlock.hash, nil
+	}
+	scanner.relevantBlock = nil // clear so we don't keep checking if it is a mainchain block
+	return nil, nil
+}
+
+// trackExternalTx records the script associated with a tx to enable spv
+// wallets locate the tx in a block when it is mined. Once mined, the block
+// containing the tx and the pkScripts of all of the tx outputs are recorded.
+// The recorded output scripts can then be used to subsequently check if an
+// output is spent in a mined transaction.
+func (dcr *ExchangeWallet) trackExternalTx(hash *chainhash.Hash, script []byte) {
+	defer func() {
+		dcr.log.Debugf("Script %x cached for non-wallet tx %s.", script, hash)
+	}()
+
+	tx, tracked := dcr.externalTx(hash)
+	if tracked {
+		tx.mtx.Lock()
+		tx.relevantScripts = append(tx.relevantScripts, script)
+		tx.mtx.Unlock()
+		return
+	}
+
+	dcr.externalTxMtx.Lock()
+	dcr.externalTxs[*hash] = &externalTx{
+		hash:            hash,
+		relevantScripts: [][]byte{script},
+	}
+	dcr.externalTxMtx.Unlock()
+}
+
+func (dcr *ExchangeWallet) externalTx(hash *chainhash.Hash) (*externalTx, bool) {
+	dcr.externalTxMtx.RLock()
+	tx, tracked := dcr.externalTxs[*hash]
+	dcr.externalTxMtx.RUnlock()
+	return tx, tracked
+}
+
+func (dcr *ExchangeWallet) externalTxOut(hash *chainhash.Hash, vout uint32) (*wire.TxOut, int8, bool, error) {
+	tx, tracked := dcr.externalTx(hash)
+	if !tracked {
+		return nil, 0, false, asset.CoinNotFoundError
+	}
+
+	// Lock the scanMtx to prevent attempted rescans from
+	// mutating the tx block, outputs map or tree field.
+	tx.scanMtx.RLock()
+	txBlockHash, err := tx.relevantBlockHash(dcr.getBlockHash)
+	if err != nil {
+		tx.scanMtx.RUnlock()
+		return nil, 0, false, err
+	}
+	if txBlockHash == nil {
+		tx.scanMtx.RUnlock()
+		return nil, 0, false, asset.CoinNotFoundError
+	}
+	output := tx.outputs[vout]
+	txTree := tx.tree
+	tx.scanMtx.RUnlock()
+
+	if output == nil {
+		return nil, 0, false, fmt.Errorf("tx %s has no output at index %d", hash, vout)
```

Comment:
As per my previous comment, will use a slice instead and keep all vouts gotten from the block tx. So this error message can remain valid like this. Any other error with amt or pkScript will be returned directly to avoid confusing callers and to aid troubleshooting where necessary.